open-craft/blockstore
# SE-510 [LX-129] REST API to get tag data for LabXchange

## Description

This PR concerns the creation of an API which the LabXchange frontend can query for the purpose of displaying information about items (e.g. "subject area", "subtitles language", "depth of knowledge", "license")

* Can query a list of entities at api/v1/entities/
* Tags are at api/v1/entities/<entity>/tags/
* Can be filtered by taxonomy eg tags/?taxonomy=7

It is read-only and does not have auth.

## Author Comments, Concerns, and Open Questions

One issue which came up was the use of urls for `external_id` in entities. That seemed a reasonable use case to me, but it presents an extremely challenging (to me) edge case, because many web servers do not permit url encoding of forward slashes in the path of the url, at least by default. Such that if the `external_id` is `https://url.org`, I do not know how it is possible to tell Django that you mean `api/v1/entities/https:%2F%2Furl.org/tags`.

If it's important to support urls, I will think on this issue further.

## Test Instructions

1. Create some taxonomies and tags
1. Create some entities and tag them
1. On this branch, start the server and curl the endpoints

All entities are shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/ | jq
[
  {
    "id": 3,
    "entity_type": "url",
    "external_id": "https://somethingsomething.com"
  },
  {
    "id": 1,
    "entity_type": "xblock",
    "external_id": "block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30"
  }
]
```

A single entity is shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30/ | jq
{
  "id": 1,
  "entity_type": "xblock",
  "external_id": "block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30"
}
```

All tags are shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30/tags/ | jq
{
  "tags": [
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Biochemistry"
    },
    {
      "taxonomy_uid": 9,
      "taxonomy_name": "License",
      "tag": "CC-BY-SA-4.0"
    },
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Cell biology"
    },
    {
      "taxonomy_uid": 10,
      "taxonomy_name": "Subtitles Language",
      "tag": "汉语"
    }
  ]
}
```

Filtered tags are shown:
```
$ curl -s http://0.0.0.0:18250/api/v1/entities/block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@971aadaaaa5a4886b37fee3ced711b30/tags/?taxonomies=7,9 | jq
{
  "tags": [
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Biochemistry"
    },
    {
      "taxonomy_uid": 9,
      "taxonomy_name": "License",
      "tag": "CC-BY-SA-4.0"
    },
    {
      "taxonomy_uid": 7,
      "taxonomy_name": "Subject Area",
      "tag": "Cell biology"
    }
  ]
}
```

The API is read-only:
```
$ curl -X PUT -d type=things -d external_id=stuff http://0.0.0.0:18250/api/v1/entities/
```

## TODOs
If anything isn't yet done, list it here
